### PR TITLE
feat: add support for common v2 endpoints (ping/metrics/version/config)

### PR DESCRIPTION
**Pull request by siggiskulason**
feat: add support for common v2 endpoints (ping/metrics/version/config)
    
This commit adds v2 (Ireland/Jakarta) support for four common endpoints:
- version
- metrics
- ping
- config

For information about the available commands, use `edgex-cli -h`

There are a number of ways to use the commands:
1. `edgex-cli version` displays version of each active microservice.
The version commands just shows the version string. The other commands
use a pretty-printed JSON structure (config) or a table (metrics).
 
```
$ edgex-cli version
EdgeX CLI version:  1.0.0-dev.9
core-metadata: 2.0.0-2
core-data: 2.0.0-2
core-command: 2.0.0-2

$ edgex-cli metrics
Service       CpuBusyAvg MemAlloc MemFrees MemLiveObjects MemMallocs MemSys   MemTotalAlloc
core-command  14         1572240  50446    8839           59285      75318280 11463128
core-metadata 14         1757440  56371    9918           66289      75318280 12001840
core-data     14         1691984  56145    9551           65696      75318280 11903936

$ edgex-cli config
core-metadata:
{
    "Clients": {
        "core-data": {
            "Host": "localhost",
            "Port": 59880,
            "Protocol": "http"
        },
...

```

2. `edgex-cli version -d` shows debug output - the URL used and the result

```
$ edgex-cli ping -d
core-metadata: http://localhost:59881/api/v2/ping: {"apiVersion":"v2","timestamp":"Fri Sep 10 16:17:44 BST 2021"}
core-data: http://localhost:59880/api/v2/ping: {"apiVersion":"v2","timestamp":"Fri Sep 10 16:17:44 BST 2021"}
core-command: http://localhost:59882/api/v2/ping: {"apiVersion":"v2","timestamp":"Fri Sep 10 16:17:44 BST 2021"}
core-scheduler: : Get "http://localhost:59861/api/v2/ping": dial tcp 127.0.0.1:59861: connect: connection refused
core-notification: : Get "http://localhost:59860/api/v2/ping": dial tcp 127.0.0.1:59860: connect: connection refused
device-modbus: : Get "http://localhost:59901/api/v2/ping": dial tcp 127.0.0.1:59901: connect: connection refused
device-camera: : Get "http://localhost:59985/api/v2/ping": dial tcp 127.0.0.1:59985: connect: connection refused
device-mqtt: : Get "http://localhost:59982/api/v2/ping": dial tcp 127.0.0.1:59982: connect: connection refused
device-rest: : Get "http://localhost:59986/api/v2/ping": dial tcp 127.0.0.1:59986: connect: connection refused
rules-engine: : Get "http://localhost:59701/api/v2/ping": dial tcp 127.0.0.1:59701: connect: connection refused
http-export: : Get "http://localhost:59704/api/v2/ping": dial tcp 127.0.0.1:59704: connect: connection refused
mqtt-export: : Get "http://localhost:59703/api/v2/ping": dial tcp 127.0.0.1:59703: connect: connection refused
functional-tests: : Get "http://localhost:59705/api/v2/ping": dial tcp 127.0.0.1:59705: connect: connection refused
push-to-core: : Get "http://localhost:59702/api/v2/ping": dial tcp 127.0.0.1:59702: connect: connection refused
```

3. `edgex-cli version -j` gets the raw JSON output for one service (by default core-data)

```
$ edgex-cli metrics -j
{"apiVersion":"v2","metrics":{"memAlloc":1755096,"memFrees":1008594,"memLiveObjects":9980,"memMallocs":1018574,"memSys":75580424,"memTotalAlloc":327693144,"cpuBusyAvg":11}}
```
It's possible to specify the service to use: `-m/--metadata`, `-c/--command`, `-n/--notification`, `-s/--scheduler` or `--data` (which is the default). This can also be combined with `-j` or `-d`

```
$ edgex-cli metrics -j -m
{"apiVersion":"v2","metrics":{"memAlloc":1709440,"memFrees":1008937,"memLiveObjects":9468,"memMallocs":1018405,"memSys":75580424,"memTotalAlloc":327792520,"cpuBusyAvg":10}}
```

```
$ edgex-cli ping -d -c
core-command: http://localhost:59882/api/v2/ping: {"apiVersion":"v2","timestamp":"Fri Sep 10 16:20:55 BST 2021"}
```

Help is available with
```
$ edgex-cli help
$ edgex-cli version --help
$ edgex-cli version --h
```

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-cli/blob/master/.github/Contributing.md.

## What is the current behavior?

edgex-cli does not currently support the V2 API.

## What is the new behavior?
This PR adds support in edgex-cli for the common V2 endpoints.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## New Imports

- [ ] Yes
- [x] No

## Specific Instructions
Are there any specific instructions or things that should be known prior to reviewing?

## Other information

### CODE REVIEW

File: internal/cmd/common.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+		return "core-notification"
+	} else if scheduler {
+		return "core-scheduler"
+	} else {
+		return ""
+	}
+}
+
+func dataService(cmd *cobra.Command, api string, name string) error {
+	s := config.GetCoreService("core-data")
```
Comment (lenny-intel) [inline internal/cmd/common.go:44]:
```suggestion
	s := config.GetCoreService(common.CoreDataServiceKey)
```

File: internal/cmd/common.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	value, _, _, err := format.GetServiceValue(s, api, name)
+	if err != nil {
+		cmd.Println(err.Error())
+	} else {
+		cmd.Println(value)
+	}
+	return nil
```
Comment (lenny-intel) [inline internal/cmd/common.go:51]: Why have return value if always return `nil`?

File: internal/cmd/common.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	"github.com/edgexfoundry/edgex-cli/internal/config"
+	"github.com/edgexfoundry/edgex-cli/internal/format"
+	"github.com/spf13/cobra"
+)
+
+var verbose, metadata, data, command, notification, scheduler, json bool
+
+func getSelectedService() string {
+	if metadata {
+		return "core-metadata"
```
Comment (lenny-intel) [inline internal/cmd/common.go:29]:
```suggestion
		return common.CoreMetaDataServiceKey
```
Same for other cases
Constants are here: https://github.com/edgexfoundry/go-mod-core-contracts/blob/main/common/constants.go#L256-L268

File: internal/cmd/common.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+			if err != nil {
+				cmd.Println(err.Error())
+			} else {
+				cmd.Printf("%s: %s\n", serviceName, value)
+			}
+		} else {
+			format.PrintServiceTable(api, format.Value, name)
+		}
+	}
+	return nil
```
Comment (lenny-intel) [inline internal/cmd/common.go:93]: Why have return value if always return nil?

File: internal/cmd/common.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+				cmd.Println(err.Error())
+			} else {
+				cmd.Printf("%s: %s: %s\n", serviceName, url, jsonValue)
+			}
+		} else {
+			format.PrintServiceTable(api, format.Debug, "")
+		}
+	} else if json {
+		if serviceName == "" {
+			serviceName = "core-data"
```
Comment (lenny-intel) [inline internal/cmd/common.go:71]:
```suggestion
			serviceName = common.CoreDataServiceKey
```

Comment (lenny-intel) [inline internal/cmd/common.go:71]: Please comment why handling empty service name differently for `json`

File: internal/cmd/common.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+	value, _, _, err := format.GetServiceValue(s, api, name)
+	if err != nil {
+		cmd.Println(err.Error())
+	} else {
+		cmd.Println(value)
+	}
+	return nil
+}
+
+func standardService(cmd *cobra.Command, api string, name string) error {
```
Comment (lenny-intel) [inline internal/cmd/common.go:54]: Not clear what this function's purpose is. I prefer descriptive name rather than comments.

File: internal/cmd/common.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+	} else if notification {
+		return "core-notification"
+	} else if scheduler {
+		return "core-scheduler"
+	} else {
+		return ""
+	}
+}
+
+func dataService(cmd *cobra.Command, api string, name string) error {
```
Comment (lenny-intel) [inline internal/cmd/common.go:43]: Not clear what this function's purpose is. I prefer descriptive name rather than comments.

File: internal/cmd/common.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+	cmd.Flags().BoolVarP(&verbose, "debug", "d", false, "show debug output")
+
+}
+
+func addStandardFlags(cmd *cobra.Command) {
+	addFormatFlags(cmd)
+	cmd.Flags().BoolVarP(&data, "data", "", false, "use core-data service endpoint")
+	cmd.Flags().BoolVarP(&command, "command", "c", false, "use core-command service endpoint")
+	cmd.Flags().BoolVarP(&metadata, "metadata", "m", false, "use core-metadata service endpoint")
+	cmd.Flags().BoolVarP(&scheduler, "scheduler", "s", false, "use core-scheduler service endpoint")
```
Comment (lenny-intel) [inline internal/cmd/common.go:107]:
```suggestion
	cmd.Flags().BoolVarP(&scheduler, "scheduler", "s", false, "use support-scheduler service endpoint")
```

File: internal/cmd/common.go [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+	cmd.Flags().BoolVarP(&notification, "notification", "n", false, "use core-notification service endpoint")
```
Comment (lenny-intel) [inline internal/cmd/common.go:108]:
```suggestion
	cmd.Flags().BoolVarP(&notification, "notification", "n", false, "use support-notifications service endpoint")
```

File: internal/cmd/config.go [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+import (
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/common"
+	"github.com/spf13/cobra"
+)
+
+func init() {
+	var cmd = &cobra.Command{
+		Use:   "config",
+		Short: "Returns the current configuration of EdgeX microservices",
```
Comment (lenny-intel) [inline internal/cmd/config.go:27]:
```suggestion
		Short: "Returns the current configuration of a given EdgeX microservice",
```

File: internal/cmd/metrics.go [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+import (
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/common"
+	"github.com/spf13/cobra"
+)
+
+func init() {
+	var cmd = &cobra.Command{
+		Use:          "metrics",
+		Short:        "Optain the CPU/memory usage stats for a given service",
```
Comment (lenny-intel) [inline internal/cmd/metrics.go:27]:
```suggestion
		Short:        "Outputs the CPU/memory usage stats for a given EdgeX microservice",
```

File: internal/cmd/ping.go [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+import (
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/common"
+	"github.com/spf13/cobra"
+)
+
+func init() {
+	var cmd = &cobra.Command{
+		Use:   "ping",
+		Short: "A simple 'ping' endpoint that can be used as a service healthcheck",
```
Comment (lenny-intel) [inline internal/cmd/ping.go:27]:
```suggestion
		Short: "The ping (health check) response for a given EdgeX microservice",
```

File: internal/config/config.go [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+	return Configuration.DeviceServices[name]
+}
+
+func GetAppService(name string) service.Service {
+	return Configuration.ApplicationServices[name]
+}
+
+func init() {
+	Configuration.CoreServices = Services{
+		"core-metadata": {
```
Comment (lenny-intel) [inline internal/config/config.go:47]:
```suggestion
		common.CoreMetaDataServiceKey : {
```

Use similar constants for the others

File: internal/config/config.go [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+		"functional-tests": {
+			Host: "localhost",
+			Port: 59705,
+		},
+		"push-to-core": {
+			Host: "localhost",
+			Port: 59702,
+		},
+	}
+}
```
Comment (lenny-intel) [inline internal/config/config.go:115]: If we are supporting these the `version` command needs to be specific to each service since they can be release on a different cadence. For example the core/support service are currently 2.0.0, but `device-coap` and `app-service-configurable` are at 2.0.1 version.

Comment (lenny-intel) [reply internal/config/config.go:115]: IMO we need some way to address app service  and device service generically rather than have the above list. Obviously the port would be a required parameter to the command. thinking out loud... Maybe for the `service-name` we accept `device` and `app` or `device:<port>` and `app:<port>`.

File: internal/format/format.go [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+func PrintServices(service config.Services, endpoint string, ft FormatType, jsonValueName string) {
+	for serviceName, service := range service {
+
+		var value, jsonValue, url string
+		var err error
+		value, jsonValue, url, err = GetServiceValue(service, endpoint, jsonValueName)
+		if err != nil {
+			if ft == Debug {
+				fmt.Printf("%s: %s: %s\n", serviceName, url, err.Error())
```
Comment (lenny-intel) [inline internal/format/format.go:51]: Why only printing the error when in DEBUG?? Is Debug really Verbose?

File: internal/format/format.go [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+	"github.com/edgexfoundry/edgex-cli/internal/config"
+	"github.com/edgexfoundry/edgex-cli/internal/service"
+)
+
+type FormatType int
+
+const (
+	Debug FormatType = iota
+	JSON
+	Value
```
Comment (lenny-intel) [inline internal/format/format.go:33]: What is the difference and why do we need both?

File: internal/format/format.go [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+					fmt.Printf("%s: %s\n", serviceName, value)
+				}
+			default:
+				fmt.Println("Unknown type")
+			}
+		}
+	}
+}
+
+func GetServiceValue(s service.Service, endpoint string, jsonValueName string) (value string, jsonStr string, url string, err error) {
```
Comment (lenny-intel) [inline internal/format/format.go:74]: All public funcs/structs should have Godoc style comments describing the func/struct

File: internal/service/service.go [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+	Host string
+	Port int
+}
+
+func (c Service) EndpointURL(endpoint string) string {
+	url := fmt.Sprintf("http://%s:%v%s", c.Host, c.Port, endpoint)
+	return url
+}
+
+func (c Service) Endpoint(endpoint string) (string, string, error) {
```
Comment (lenny-intel) [inline internal/service/service.go:35]: Please use better descriptive names. Like `CallEndpoint` for this one.

File: internal/service/service.go [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+	"io/ioutil"
+	"net/http"
+)
+
+type Service struct {
+	Host string
+	Port int
+}
+
+func (c Service) EndpointURL(endpoint string) string {
```
Comment (lenny-intel) [inline internal/service/service.go:30]: Please use better descriptive names. Like `GetEndpointUrl` for this one.

File: internal/service/service.go [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+
+package service
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+)
+
+type Service struct {
```
Comment (lenny-intel) [inline internal/service/service.go:25]: All public funcs/structs should have Godoc style comments describing the func/struct

File: internal/format/format.go (hunk 17 above)
Comment (lenny-intel) [reply internal/format/format.go:74]: Also, maybe a more descriptive function name. From the code , I can't determine the purpose of the function.

File: snap/local/runtime-helpers/bin/edgex-cli-wrapper.sh [hunk 21]
```diff
@@ -1,10 +1,10 @@
...
-    cp "$SNAP/res/sample-configuration.toml" "$HOME/.edgex-cli/configuration.toml"
-    echo "Created $HOME/.edgex-cli/configuration.toml from sample configuration file."
-fi
+#if [ ! -f "$HOME/.edgex-cli/configuration.toml" ]; then
+#    mkdir -p "$HOME/.edgex-cli"
+#    cp "$SNAP/res/sample-configuration.toml" "$HOME/.edgex-cli/configuration.toml"
+#    echo "Created $HOME/.edgex-cli/configuration.toml from sample configuration file."
+#fi
 
 exec "$@"
```
Comment (lenny-intel) [inline snap/local/runtime-helpers/bin/edgex-cli-wrapper.sh:10]: Is this file needed. If so, remove commented out code..

File: README.md [hunk 22]
```diff
@@ -5,225 +5,3 @@
 
 A command line interface to interact with EdgeX microservices. Replaces the need to manually construct complex CURL commands and/or maintain developer scripts.
```
Comment (lenny-intel) [inline README.md:6]: I think we need a placeholder link to the on-line docs, which should cover how to install, run, etc.

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: Makefile [hunk 23]
```diff
@@ -15,14 +15,16 @@ ifndef GOBIN
 endif
 
 BINARY=edgex-cli
-
-VERSION=$(shell cat ./VERSION 2>/dev/null || echo 0.0.0)
-GOFLAGS=-ldflags "-X github.com/edgexfoundry/edgex-cli/cmd/version.Version=$(VERSION)"
+ 
+VERSION=$(shell git describe --tags --abbrev=0 | sed 's/v//' 2>/dev/null || echo 0.0.0)
+TIME=$(shell date)
+GOFLAGS=-ldflags "-X 'github.com/edgexfoundry/edgex-cli.BuildVersion=$(VERSION)' -X 'github.com/edgexfoundry/edgex-cli.BuildTime=$(TIME)'"
```
[the remark this answers is not in the document]
Comment (ernestojeda) [reply Makefile:21]: I think this is not correct. For instance, if this were run today, `git describe --tags --abbrev=0 | sed 's/v//'` would return `2.0.0-dev.19` this is the previously tagged version which already exists thus giving the wrong version. The correct version number would be `2.0.0-dev.20` which can be found here: https://github.com/edgexfoundry/edgex-cli/blob/semver/main#L1. This is just an artifact of how git-semver works. We are working to rectify this as part of Kamakura with our explore into the use of `semantic-release`.

We should probably stick with `VERSION=$(shell cat ./VERSION 2>/dev/null || echo 0.0.0)` like we do for every other repository.

[the remark this answers is not in the document]
Comment (lenny-intel) [reply Makefile:21]: @ernestojeda , Thanks for the feedback!

[the remark this answers is not in the document]
Comment (siggiskulason) [reply Makefile:21]: @lenny-intel / @ernestojeda - thanks for this comment. I'll change it to be inline with all the other repositories. This logic of using the git tag came from the way we do this in the snaps - and I suppose that highlights the two different scenarios we have. In the Docker/Jenkins pipeline build we do the build before the commit is tagged - as tagging is the last step in the build- so as you rightly point out, we should use the VERSION file generated using the git-semver process. However - when building the snap, we do that build after the Jenkins build process has run, the commit tagged and the semver bumped to the next version - so the correct thing there would be to use the latest tag.

File: Makefile [hunk 24]
```diff
@@ -15,14 +15,16 @@ ifndef GOBIN
...
 ARTIFACT_ROOT?=bin
 
 build:
 	@echo "GOPATH=$(GOPATH)"
-	$(GO) build -o $(BINARY) $(GOFLAGS)
+	go mod tidy
```
[the remark this answers is not in the document]
Comment (siggiskulason) [reply Makefile:26]: done

File: Makefile [hunk 25]
```diff
@@ -49,9 +51,9 @@ test:
 install:
 	@echo "GOBIN=$(GOBIN)"
 	$(GO) install $(GOFLAGS)
-	mkdir -p $(HOME)/.edgex-cli
-	cp ./res/sample-configuration.toml $(HOME)/.edgex-cli/configuration.toml
-	@echo "Configuration file $(HOME)/.edgex-cli/configuration.toml created"
+#	mkdir -p $(HOME)/.edgex-cli
+#	cp ./res/sample-configuration.toml $(HOME)/.edgex-cli/configuration.toml
+#	@echo "Configuration file $(HOME)/.edgex-cli/configuration.toml created"
```
[the remark this answers is not in the document]
Comment (siggiskulason) [reply Makefile:56]: done

File: internal/config/config.go (hunk 14 above)
Comment (siggiskulason) [reply internal/config/config.go:115]: I've removed these for now and will focus on supporting the core services. We can add a generic way of supporting these services at a later stage as you suggest.

File: internal/format/format.go (hunk 15 above)
Comment (siggiskulason) [reply internal/format/format.go:51]: Yes, as per line 34 of the ADR draft:

```
`-d`, `--debug` | show additional output for debugging purposes (e.g. REST URL, request JSON, …). This command-line arg will replace -v, --verbose and will no longer trigger output of the response JSON (see -j, --json).
```
I could rename the constant as `Verbose` if that reflects the intention better?

File: internal/format/format.go (hunk 16 above)
Comment (siggiskulason) [reply internal/format/format.go:33]: `version` and `ping` just need to print a single value. `metrics` returns a list of strings and `config` a hierarchy. This is really there to make the output of `version` and `ping` a bit more user-friendly and just display that single value, rather than the whole JSON output. See `edgex-cli version` vs `edgex-cli metrics`

File: internal/format/format.go (hunk 17 above)
Comment (siggiskulason) [reply internal/format/format.go:74]: Good point! I'll update all of them.

File: internal/cmd/common.go (hunk 1 above)
Comment (siggiskulason) [reply internal/cmd/common.go:44]: done

File: internal/cmd/common.go (hunk 2 above)
Comment (siggiskulason) [reply internal/cmd/common.go:51]: This method should not even be there and I've now removed it.

File: internal/cmd/common.go (hunk 3 above)
Comment (siggiskulason) [reply internal/cmd/common.go:29]: thanks. all updated now.

File: internal/cmd/common.go (hunk 4 above)
Comment (siggiskulason) [reply internal/cmd/common.go:93]: Absolutely. As I'm not returning the error value but consuming it, I removed the error result value and this return statement.

File: internal/cmd/common.go (hunk 5 above)
Comment (siggiskulason) [reply internal/cmd/common.go:71]: updated to use CoreMetadata

Comment (siggiskulason) [reply internal/cmd/common.go:71]: Added:

```
// In the case of JSON output, we only print the result from a single service,
// core-metadata by default
```

File: internal/cmd/common.go (hunk 6 above)
Comment (siggiskulason) [reply internal/cmd/common.go:54]: changed to 

```
// commonServiceEndpointImplementation is used internally by config/metrics/ping/version
// to implement the shared functionality of those commands 
func commonServiceEndpointImplementation(cmd *cobra.Command, api string, name string) {
```

File: internal/cmd/common.go (hunk 7 above)
Comment (siggiskulason) [reply internal/cmd/common.go:43]: It should not be in there at all. removed.

File: internal/cmd/common.go (hunk 8 above)
Comment (siggiskulason) [reply internal/cmd/common.go:107]: oops, absolutely, copy-and-paste error there in the naming of those services. well spotted. thanks. fixed.

File: internal/cmd/common.go (hunk 9 above)
Comment (siggiskulason) [reply internal/cmd/common.go:108]: as above

File: internal/cmd/config.go (hunk 10 above)
Comment (siggiskulason) [reply internal/cmd/config.go:27]: well, it doesn't really. It returns the configuration of all core services, unless you specifically pick the one to use.

File: internal/cmd/metrics.go (hunk 11 above)
Comment (siggiskulason) [reply internal/cmd/metrics.go:27]: I changed this to  `Outputs the CPU/memory usage stats for EdgeX microservices`

File: internal/cmd/ping.go (hunk 12 above)
Comment (siggiskulason) [reply internal/cmd/ping.go:27]: As above, set this to `The ping (health check) response for EdgeX microservices`

File: internal/service/service.go (hunk 18 above)
Comment (siggiskulason) [reply internal/service/service.go:35]: done

File: internal/service/service.go (hunk 19 above)
Comment (siggiskulason) [reply internal/service/service.go:30]: good point. done.

File: internal/service/service.go (hunk 20 above)
Comment (siggiskulason) [reply internal/service/service.go:25]: absolutely. added.

File: snap/local/runtime-helpers/bin/edgex-cli-wrapper.sh (hunk 21 above)
Comment (siggiskulason) [reply snap/local/runtime-helpers/bin/edgex-cli-wrapper.sh:10]: not anymore. removed.

File: README.md (hunk 22 above)
Comment (siggiskulason) [reply README.md:6]: yes, this README needs to be updated in a separate PR once we have the on-line docs ready.

**Comment by siggiskulason** [comment]
Thank you very much @lenny-intel for the helpful comments. I have responded to them inline, updated the code, force-pushed an update and marked the comments as resolved.

File: Makefile [hunk 26]
```diff
@@ -15,23 +15,27 @@ ifndef GOBIN
...
 
 # initial impl. Feel free to override. Please keep ARTIFACT_ROOT coming from env though. CI/CD pipeline relies on this
 build-all:
 	@echo "GOPATH=$(GOPATH)"
 
-	GOOS=linux GOARCH=amd64 $(GO) build -o ${ARTIFACT_ROOT}/$(BINARY)-linux-amd64 $(GOFLAGS)
-	GOOS=linux GOARCH=arm64 $(GO) build -o ${ARTIFACT_ROOT}/$(BINARY)-linux-arm64 $(GOFLAGS)
-	GOOS=darwin GOARCH=amd64 $(GO) build -o ${ARTIFACT_ROOT}/$(BINARY)-mac $(GOFLAGS)
-	GOOS=windows GOARCH=amd64 $(GO) build -o ${ARTIFACT_ROOT}/$(BINARY).exe $(GOFLAGS)
+	GOOS=linux GOARCH=amd64 $(GO) build -o ${ARTIFACT_ROOT}/$(BINARY)-linux-amd64 $(GOFLAGS) ./cmd/edgex
```
Comment (lenny-intel) [inline Makefile:35]: binary is `edgex`, not `edgex-cli`

File: internal/cmd/common.go [hunk 27]
```diff
@@ -0,0 +1,104 @@
...
+
+}
+
+func addStandardFlags(cmd *cobra.Command) {
+	addFormatFlags(cmd)
+	cmd.Flags().BoolVarP(&data, "data", "", false, "use core-data service endpoint")
+	cmd.Flags().BoolVarP(&command, "command", "c", false, "use core-command service endpoint")
+	cmd.Flags().BoolVarP(&metadata, "metadata", "m", false, "use core-metadata service endpoint")
+	cmd.Flags().BoolVarP(&scheduler, "scheduler", "s", false, "use support-scheduler service endpoint")
+	cmd.Flags().BoolVarP(&notification, "notification", "n", false, "use support-notification service endpoint")
```
Comment (lenny-intel) [inline internal/cmd/common.go:102]:
```suggestion
	cmd.Flags().BoolVarP(&notifications, "notifications", "n", false, "use support-notifications service endpoint")
```
since the service's actual name is support-notifications

File: internal/cmd/config.go (hunk 10 above)
Comment (lenny-intel) [inline internal/cmd/config.go:27]:
```suggestion
		Short: "Returns the current configuration of all EdgeX core/support microservices",
```

File: internal/cmd/metrics.go [hunk 28]
```diff
@@ -0,0 +1,38 @@
...
+
+import (
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/common"
+	"github.com/spf13/cobra"
+)
+
+func init() {
+	var cmd = &cobra.Command{
+		Use:          "metrics",
+		Short:        "Outputs the CPU/memory usage stats for EdgeX microservices",
```
Comment (lenny-intel) [inline internal/cmd/metrics.go:27]:
```suggestion
		Short:        "Outputs the CPU/memory usage stats for all EdgeX core/support microservices",
```

File: internal/cmd/ping.go [hunk 29]
```diff
@@ -0,0 +1,39 @@
...
+
+import (
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/common"
+	"github.com/spf13/cobra"
+)
+
+func init() {
+	var cmd = &cobra.Command{
+		Use:   "ping",
+		Short: "The ping (health check) response for EdgeX microservices",
```
Comment (lenny-intel) [inline internal/cmd/ping.go:27]:
```suggestion
		Short: "The ping (health check) response for all EdgeX core/support microservices",
```

File: internal/format/format.go (hunk 15 above)
Comment (lenny-intel) [reply internal/format/format.go:38]: I like it!

File: internal/format/format.go [hunk 30]
```diff
@@ -0,0 +1,76 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0'
+ */
+
+package format
```
Comment (lenny-intel) [inline internal/format/format.go:17]: This package and code is making the actual REST calls, correct? If so, `format` doesn't seem the best name. I was expecting just utilities to format data. Also not using the clients in go-mod-core-contracts instead of the direct REST calls?

File: snap/local/runtime-helpers/bin/edgex-cli-wrapper.sh [hunk 31]
```diff
@@ -1,10 +1,3 @@
 #!/bin/sh -e
 
-# configuration.toml is required for the client to run
-if [ ! -f "$HOME/.edgex-cli/configuration.toml" ]; then
-    mkdir -p "$HOME/.edgex-cli"
-    cp "$SNAP/res/sample-configuration.toml" "$HOME/.edgex-cli/configuration.toml"
-    echo "Created $HOME/.edgex-cli/configuration.toml from sample configuration file."
-fi
-
 exec "$@"
```
Comment (lenny-intel) [inline snap/local/runtime-helpers/bin/edgex-cli-wrapper.sh:3]: Is this wrapper still needed? Seem like it no longer adds any value.

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: Makefile (hunk 26 above)
Comment (siggiskulason) [reply Makefile:35]: BINARY is set to `edgex-cli` in line 17. However, this refers to the directory containing `/cmd/edgex/main.go`. That directory should really be named `edgex-cli` as well for consistency, so I've renamed it and updated this accordingly. Everything should now therefore use `edgex-cli`

File: internal/cmd/common.go (hunk 27 above)
Comment (siggiskulason) [reply internal/cmd/common.go:102]: thanks. fixed.

File: internal/cmd/config.go (hunk 10 above)
Comment (siggiskulason) [reply internal/cmd/config.go:27]: fixed

File: internal/cmd/metrics.go (hunk 28 above)
Comment (siggiskulason) [reply internal/cmd/metrics.go:27]: fixed

File: internal/cmd/ping.go (hunk 29 above)
Comment (siggiskulason) [reply internal/cmd/ping.go:27]: fixed

File: internal/format/format.go (hunk 30 above)
Comment (siggiskulason) [reply internal/format/format.go:17]: good point. I created this package originally with the view in mind that I might need some code to format the data as I implement the methods. However, as you pointed out, it should be using the go-mod-core-contracts implementation instead anyway. I'm replacing this now with a different implementation using go-mod-core-contracts and tabular output and will check that in later today.

File: snap/local/runtime-helpers/bin/edgex-cli-wrapper.sh (hunk 31 above)
Comment (siggiskulason) [reply snap/local/runtime-helpers/bin/edgex-cli-wrapper.sh:3]: Not really, but it might be needed for the new commands to be added. If not, then I suggest removing it in the next PR.

**Comment by siggiskulason** [comment]
Hi @lenny-intel - thanks again for your review. I have responded to all your comments and made some changes - I removed the format package and also for `metrics` added the SDK method to get the metrics - which I then print out in a table, rather than printing out the JSON result. I still print out the config as a JSON for now at least, as that can have an hierarchical structure - and I just get the raw JSON for version/ping. For the other methods I'm now starting to add I'm using the SDK functions. I've force-pushed the changes and marked your comments as resolved.

**Review by lenny-intel — APPROVED** [review]
LGTM
